pkgs/by-name/so/sourcegit/package.nix

```diff
@@ -0,0 +1,112 @@
+{
+  lib,
+  stdenv,
+  buildDotnetModule,
+  fetchFromGitHub,
+  dotnetCorePackages,
+  copyDesktopItems,
+  makeDesktopItem,
+  libicns,
+
+  libXcursor,
+  libXext,
+  libXi,
+  libXrandr,
+
+  git,
+  xdg-utils,
+}:
+
+buildDotnetModule (finalAttrs: {
+  pname = "sourcegit";
+  version = "2025.01";
+
+  src = fetchFromGitHub {
+    owner = "sourcegit-scm";
+    repo = "sourcegit";
+    tag = "v${finalAttrs.version}";
+    hash = "sha256-6FO0WRUcWm7xnuw6Br97+cWZhvzIOesg/eANZzeyxZo=";
+  };
+
+  patches = [ ./fix-darwin-git-path.patch ];
+
+  dotnet-sdk = dotnetCorePackages.sdk_9_0;
+  dotnet-runtime = dotnetCorePackages.runtime_9_0;
+
+  nugetDeps = ./deps.json;
+
+  projectFile = [ "src/SourceGit.csproj" ];
+
+  executables = [ "SourceGit" ];
+
+  dotnetFlags = [
+    "-p:DisableUpdateDetection=true"
+    "-p:DisableAOT=true"
+  ];
+
+  nativeBuildInputs = lib.optionals stdenv.hostPlatform.isLinux [
+    copyDesktopItems
+    libicns
+  ];
+
+  # these are dlopen-ed at runtime
+  # libXi is needed for right-click support
+  # not sure about what the other ones are needed for, but I'll include them anyways
+  runtimeDeps = [
+    libXcursor
+    libXext
+    libXi
+    libXrandr
+  ];
+
+  desktopItems = [
+    (makeDesktopItem {
+      name = "SourceGit";
+      exec = "SourceGit";
+      icon = "SourceGit";
+      desktopName = "SourceGit";
+      terminal = false;
+      comment = finalAttrs.meta.description;
+    })
+  ];
+
+  # add fallback binaries to use if the user doesn't already have them in their PATH
+  preInstall = ''
+    makeWrapperArgs+=(
+      --suffix PATH : ${
+        lib.makeBinPath [
+          git
+          xdg-utils
+        ]
+      }
+    )
+  '';
+
+  postInstall =
+    lib.optionalString stdenv.hostPlatform.isLinux ''
+      icns2png -x build/resources/App.icns
+      for f in App_*x32.png; do
+          res=''${f//App_}
+          res=''${res//x32.png}
+          install -Dm644 $f "$out/share/icons/hicolor/$res/apps/SourceGit.png"
+      done
+    ''
+    + lib.optionalString stdenv.hostPlatform.isDarwin ''
+      install -Dm644 build/resources/app/App.icns $out/Applications/SourceGit.app/Contents/Resources/App.icns
+
+      substitute build/resources/app/App.plist $out/Applications/SourceGit.app/Contents/Info.plist \
+        --replace-fail "SOURCE_GIT_VERSION" "${finalAttrs.version}"
+
+      mkdir -p $out/Applications/SourceGit.app/Contents/MacOS
+      ln -s $out/bin/SourceGit $out/Applications/SourceGit.app/Contents/MacOS/SourceGit
+    '';
+
+  meta = {
+    changelog = "https://github.com/sourcegit-scm/sourcegit/releases/tag/${finalAttrs.src.tag}";
```

Comment:
This would break if someone overrides the package with a `src` that doesn't have a `tag`

Maybe that's not an issue, because `meta.changelog` is probably only read for in-tree packages, and is likely unused for out-of-tree overrides.

I think using `"v${finalAttrs.version}"` would be ok though.

Reply:
The same could be said for using `src.rev` as well, and that's pretty common

IMO it's fine, since we're not evaling `meta` anyways.

Reply:
> The same could be said for using `src.rev` as well, and that's pretty common

I've normally seen that done _without_ using `finalAttrs`, so it wouldn't be affected by overrides.

I haven't managed to find any guidance on when we should use fixed vs overridable references to attrs.

Instead, I usually see strong opinions banded around without any real explanation or justification 😥

Reply:
Yeah, working fixed-point packages is a bit annoying, especially when dealing with FOD hashes not being invalidated (because of the default `name = "source"`).

I don't know how much fixed-point packages worsen the eval performance, but I like them otherwise.

The main issue is that we are in a state inbetween totally fixed-point package definitions and (mostly) static package definitions. And new packages are being written in both styles every day.

It doesn't help that most wrappers for mkDerivation ATM don't support passing in a fixed-point attrset.